### PR TITLE
FS-1987: Add doc-strings for applicants response

### PR DESCRIPTION
**Pull request by tferns**
- Add doc-strings to methods & change terminology to be consistent throughout app
- Handle one unaccounted for use-case where grouped_fields has empty, None or missing answers

### CODE REVIEW

File: app/assess/models/ui/applicants_response.py [hunk 1]
```diff
@@ -129,6 +130,27 @@ def from_dict(cls, data: dict):
...
+            - "multilineTextField": A multi-line text field.
+            - "websiteField": A website field.
+            - "datePartsField": A date field.
+            - "emailAddressField": An email address field.
+            - "telephoneNumberField": A telephone number field.
+        - answer (str): The answer for the UI component.
+
+    :return (ApplicantResponseComponent): An instance of the UI component class.
+    - An instance of a subclass of UIComponent, representing the UI component for the given item.
+    """
```
Comment (harryyo) [inline app/assess/models/ui/applicants_response.py:153]: Nicely laid out 👍

**Review by harryyo — CHANGES_REQUESTED** [review]
Hi Thomas, nice stuff. I've noted where a couple are perhaps a bit wordy. Could they be simplified maybe. Have pointed out the ones that seem great length 👍

File: app/assess/models/ui/applicants_response.py [hunk 2]
```diff
@@ -187,6 +209,40 @@ def _ui_component_from_factory(item: dict):
...
+    "question" key with a tuple of the heading text and the description question
+    text, a "field_id" key with the field ID of the heading, and an "answer" key
+    with a list of tuples of descriptions and corresponding amounts. The
+    "presentation_type" key will be set to "grouped_fields". The returned set of
+    field IDs will include the field IDs of all headings in the response.
+
+    If a description or amount is missing for a heading, the heading will be
+    added to the returned list as a regular field with the "question" key set to
+    the heading text and the "field_id" key set to the field ID of the heading.
+    The "presentation_type" key will be set to "heading".
```
[the remark this answers is not in the document]
Comment (tferns) [reply app/assess/models/ui/applicants_response.py]: I've changed this to have comments in the function itself... its mostly just repeating what the code does

File: app/assess/models/ui/applicants_response.py [hunk 3]
```diff
@@ -262,15 +336,35 @@ def _get_item_by_presentation_type_index(response, presentation_type, index):
 def _convert_checkbox_items(
     response: Iterable[dict],
 ) -> tuple[list[dict], set[str]]:
-    field_ids = {
+    """
+    Convert checkbox items in the response into a tuple of a list of dictionary
+     items and a set of field IDs.
+
+    This method processes checkbox items in the response and returns a tuple of
+    a list of dictionary items and a set of field IDs. The list of dictionary
```
[the remark this answers is not in the document]
Comment (tferns) [reply app/assess/models/ui/applicants_response.py]: Good shout have removed

File: app/assess/models/ui/applicants_response.py [hunk 4]
```diff
@@ -312,6 +412,36 @@ def _flatten_field_ids(field_id):
...
+        field_type: a string representing the type of field, which should not be "numberField" for the fields that this
+        function processes
+        question: a list of strings representing the questions for the fields
+        field_id: a list of strings or lists of strings representing the field IDs for the fields
+        answer: a list of tuples, where each tuple contains a string representing the question for the field and a string
+        representing the answer for the field
+        This function processes the fields in the input response that have a presentation_type of "grouped_fields" and a
+        field_type other than "numberField". It converts each field into a dictionary representing a single text field and
+        appends it to the list of text items. The field IDs for the processed fields are added to the set of processed field
+        IDs.
```
[the remark this answers is not in the document]
Comment (tferns) [reply app/assess/models/ui/applicants_response.py]: Black doesn't wrap these lines automatically so have to do it manually, have tried to fix it up a bit

Flake8 also just warns and doesn't auto fix so that's a tad annoying

**Comment by tferns** [comment]
> Hi Thomas, nice stuff. I've noted where a couple are perhaps a bit wordy. Could they be simplified maybe. Have pointed out the ones that seem great length 👍

Thanks for the feedback, have pushed https://github.com/communitiesuk/funding-service-design-assessment/pull/74/commits/b89eabf1265641a30a0bf317c61ccb2ab4a05ab5 to address some of your comments

File: app/assess/models/ui/applicants_response.py [hunk 5]
```diff
@@ -187,15 +210,28 @@ def _ui_component_from_factory(item: dict):
 def _convert_heading_description_amount(
     response: Iterable[dict],
 ) -> tuple[list[dict], set[str]]:
+    """Convert a list of dictionaries representing a respose question and
```
Comment (harryyo) [inline app/assess/models/ui/applicants_response.py:213]: typo *response

**Review by harryyo — APPROVED** [review]
Fantastic - v. clear 👍

File: app/assess/models/ui/applicants_response.py (hunk 5 above)
Comment (tferns) [reply app/assess/models/ui/applicants_response.py]: Doh, good spot, thanks https://github.com/communitiesuk/funding-service-design-assessment/pull/74/commits/b0a0995b0ee14e4b0065d55b4ac17febed38a555
